text/0000-task-builder.md

```diff
@@ -0,0 +1,200 @@
+    let n = cx.argument::<JsNumber>(0)?.value();
+    let cb = cx.argument::<JsFunction>(1)?;
+
+    cx.task(move || {
+            let result = n + 1.0;
+
+            move |mut cx| Ok(cx.number(result))
+        })
+        .schedule(cb)
+}
+
+register_module!(mut cx, {
+    cx.export_function("add_one", addOne)
+});
+```
+
+The `add_one` method can be called from Javascript:
+
+```js
+const { addOne } from './native';
+
+addOne(41, (err, val) => {
+    if (err) {
+        console.error(err);
+    } else {
+        console.log(`Result: ${val}`);
+    }
+});
+```
+
+The `cx.task(...)` method creates a new `TaskBuilder`. It expects a single argument. This argument is a `closure` that will be executed on the `libuv` threadpool.
+
+Any setup that requires the V8 VM (e.g., getting an `f64` from a `Number`) must be performed outside of this closure.
+
+The closure is expected to return _another closure_. This higher-order method will be familiar to many Javascript developers. The returned closure is the `complete` method and will be performed back on the main Javascript thread. In the `complete` closure, the user may convert Rust types back to Javascript types.
+
+Finally, a `schedule` method is provided to submit the task to the queue. For conveinance, it returns `JsUndefined`, matching Javascript semantics for most asynchronous methods.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to low-level Task
+
+The existing low-level `Task` implementation in C++ is modified to match a new ownership model. The `perform` model will _consume_ self. The `complete` method must be changed to no longer pass `self` to prevent use-after-free.
+
+```rust
+pub fn Neon_Task_Schedule(task: *mut c_void,
+                          perform: unsafe extern fn(*mut c_void) -> *mut c_void,
+                          complete: unsafe extern fn(*mut c_void, &mut Local),
+                          callback: Local);
+```
+
+### Private `BaseTask` trait
+
+A new (private) `BaseTask` trait is introduced. The `BaseTask` trait is very similar to the existing `Task` trait except `fn perform` is changed to take ownership of `self` and `fn complete` is changed to an associated method (does not use self).
+
+A single `Output` parameter is provided and it is up to the user to return a `Result` if desired.
+
+```rust
+trait BaseTask: Send + Sized + 'static {
+    type Output: Send + 'static;
+    type JsEvent: Value;
+
+    fn perform(self) -> Self::Output;
+    fn complete<'a>(cx: TaskContext<'a>, result: Self::Output) -> JsResult<Self::JsEvent>;
+    fn schedule(self, callback: Handle<JsFunction>);
+}
+```
+
+### `Task` trait
+
+An implementation of `BaseTask` is added for `T: Task` to provide backwards compatibility to existing implementations of `Task`. Since `complete` is an associated method, `self` is threaded through the `Output` parameter.
+
+The existing `Task` trait will be _deprecated_.
```

Comment:
Let's mention why (it's less convenient and less powerful) and also that we won't remove it until Neon users have had enough time to migrate to the new API.